lib/Dialect/Ibis/Transforms/IbisInlineSBlocksPass.cpp

```diff
@@ -0,0 +1,146 @@
+    Block *sblockBody = op.getBodyBlock();
+    sblockBody->moveBefore(succBlock);
+    BlockReturnOp ret = cast<BlockReturnOp>(sblockBody->getTerminator());
+
+    // Replace the ibis.sblock return values with the values that were defined
+    // (returned from) within the sblock body, and erase the return op.
+    for (auto [res, val] : llvm::zip(op.getResults(), ret.getRetValues()))
+      res.replaceAllUsesWith(val);
+    ret.erase();
+
+    // Create control flow from the predecessor block to the sblock body.
+    if (getNumOps(predBlock) == 1) {
+      // Predecessor block only contained the sblock, so we can merge it
+      // with the sblock body.
+      predBlock->getOperations().splice(predBlock->end(),
+                                        sblockBody->getOperations());
+      sblockBody->erase();
+      sblockBody = predBlock;
+
+      // Outgoing control flow should start from the end of the merged block.
+      rewriter.setInsertionPointToEnd(predBlock);
+    } else {
+      rewriter.setInsertionPointToEnd(predBlock);
+      rewriter.create<cf::BranchOp>(loc, sblockBody);
+      // Outgoing control flow should start from the end of the sblock body.
+      rewriter.setInsertionPointToEnd(sblockBody);
+    }
+
+    // Create control flow from the sblock body to the successor block.
+    if (getNumOps(succBlock) == 1) {
+      // Successor block only contains a branch, so we can merge it with the
+      // sblock body.
+      sblockBody->getOperations().splice(sblockBody->end(),
+                                         succBlock->getOperations());
+      rewriter.eraseBlock(succBlock);
+    } else {
+      // Successor block contains more than just a branch, so we need to create
+      // a branch to it.
+      rewriter.create<cf::BranchOp>(loc, succBlock);
+    }
+
+    // Add an attribute for the source block location.
+    llvm::SmallVector<NamedAttribute> blockAttrs;
+    blockAttrs.push_back(
+        rewriter.getNamedAttr("loc", static_cast<LocationAttr>(loc)));
+    if (auto maxThreads = op.getMaxThreadsAttr())
+      blockAttrs.push_back(rewriter.getNamedAttr("maxThreads", maxThreads));
+
+    blockAttrMap[sblockBody] = rewriter.getDictionaryAttr(blockAttrs);
+    rewriter.eraseOp(op);
+    return success();
+  }
+
+  DenseMap<Block *, DictionaryAttr> &blockAttrMap;
+};
+
+} // anonymous namespace
+
+void InlineSBlocksPass::runOnOperation() {
+  MethodOp parent = getOperation();
+  ConversionTarget target(getContext());
+  target.addLegalDialect<cf::ControlFlowDialect>();
+  target.addIllegalOp<ibis::StaticBlockOp>();
+
+  DenseMap<Block *, DictionaryAttr> blockAttrMap;
+  RewritePatternSet patterns(&getContext());
+  patterns.add<InlineSBlocksPattern>(&getContext(), blockAttrMap);
+
+  if (failed(applyPartialConversion(parent, target, std::move(patterns))))
+    return signalPassFailure();
+
+  // Add block annotations to the parent method op.
+  llvm::SmallVector<NamedAttribute> blockInfo;
```

Comment:
As we discussed in our 1:1 today, do you want to implement this with 'guard' instructions instead?

Reply:
I would prefer that. It seems less fragile than tracking with attrs in the parent.

Reply:
Agreed - guard ops with side-effects to avoid reordering seems to be the better way to go -requires this PR to be a bit more reach seeing as the reblocking pass needs to factor that in as well.